Gp 1.3.1 updates

jenswi-linaro opened the pull request:
This PR upgrades from TEE Internal Core API version 1.1 to 1.3.1.

Note that we're skipping all the versions in between. This is not just because I'm lazy, it's also because it seems unnecesary complicated to take it in small steps. For example, the size_t to uint32_t change from 1.0 to 1.1 is undone partially from 1.1 to 1.1.1, a few more things are done in 1.2 and 1.2.1 but it's not completed until 1.3.

There are still a few things left before OP-TEE is fully 1.3.1 compliant. Here's a list of things I've noted so far:

4.9.4 Operation Parameters in the Internal Client API

The implementation SHALL enforce the following restrictions on
TEE_PARAM_TYPE_MEMREF_XXX values:
• Where all or part of the referenced memory buffer was passed to the TA
  from the REE or from another TA, the implementation SHALL NOT result in
  downgrade of the security characteristics of the buffer – see Table 4-5.
• Where all or part of the referenced buffer was allocated by the TA with
  the TEE_MALLOC_NO_SHARE hint, the implementation SHALL raise a Panic for
  the TA.

****** Note: OP-TEE is not compliant with the above ******

For TEE_OpenTASession() and TEE_InvokeTACommand(), OP-TEE doesn't handle the TEE_ERROR_CANCEL specifically. All possible error code may need to be double checked.

For TEE_SetOperationKey() and TEE_SetOperationKey2(), OP-TEE will panic the TA unless the operation is in initial state. This isn't correct any longer, the operation state should instead be reset to initial state.

SHA-3 and TEE_DigestExtract() is missing. Will implement it on top of _utee_hash_final() as an alternative way of finalizing a hash.

TEE_AEEncryptFinal(), OP-TEE doesn't treat destData as optional, this should be fixed.

TEE_AsymmetricEncrypt(), TEE_AsymmetricDecrypt(), TEE_ERROR_CIPHERTEXT_INVALID isn't used properly by OP-TEE for all algorithms, this should be fixed.

TEE_BigIntSetBit(), TEE_BigIntAssign, TEE_BigIntAbs(), and TEE_BigIntExpMod() are missing in OP-TEE, this should be fixed.

Storage types TEE_STORAGE_PERSO and TEE_STORAGE_PROTECTED are not supported yet.

This depends on https://github.com/linaro-swg/optee_examples/pull/100 and https://github.com/OP-TEE/optee_test/pull/629

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented:
We could in the first step merge up to "Update reported version to GP Core Internal API v1.3.1" to enable merging the optee_examples and optee_test pull requests merged too.

jenswi-linaro commented:
Checkpatch complains over a few typedefs needed.

jenswi-linaro commented:
Build errors in optee_test and optee_examples are expected until [linaro-swg/optee_examples#100](https://github.com/linaro-swg/optee_examples/pull/100) and [OP-TEE/optee_test#629](https://github.com/OP-TEE/optee_test/pull/629) are merged.

etienne-lms commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -1,14 +1,103 @@
+
+/*
+ * For backwards compatibility with v1.1 as provided by OP-TEE back
+ * then, define __OPTEE_CORE_API_COMPAT_1_1 to 1.
```
do you mean line s98/100 should be the below?
```
#ifndef __OPTEE_CORE_API_COMPAT_1_1
#define __OPTEE_CORE_API_COMPAT_1_1	1
#endif
```

jenswi-linaro replied:
Please see https://github.com/OP-TEE/optee_os/pull/5689 instead.

etienne-lms commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -1,14 +1,103 @@
+ */
+#ifndef __OPTEE_CORE_API_COMPAT_1_1
+#define __OPTEE_CORE_API_COMPAT_1_1	0
+#endif
```
conflicts with the definition below.

etienne-lms reviewed (commented):
Comments for commit "Update reported version to GP Core Internal API v1.3.1".

jenswi-linaro commented:
> Comments for commit "Update reported version to GP Core Internal API v1.3.1".

I've extracted this commit into a separate PR, https://github.com/OP-TEE/optee_os/pull/5689, so we hopefully can merge it soon. I guess that CI is broken until this is merged or if we back out what was last merged to optee_test.

jenswi-linaro commented on lib/libutee/tee_api_operations.c:
```diff
@@ -68,20 +68,23 @@ TEE_Result TEE_AllocateOperation(TEE_OperationHandle *operation,
 
-	case TEE_ALG_ECDSA_P192:
-	case TEE_ALG_ECDH_P192:
+	case TEE_ALG_ECDSA_SHA1:
```
`translate_compat_algo()` isn't available in libutee. Besides, I think we get fewer code paths or configuration permutations to cover when testing if we only translate in one place.

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3127,40 +3127,57 @@ TEE_Result syscall_cipher_final(unsigned long state, const void *src,
+		case TEE_ATTR_HKDF_HASH_ALGORITHM:
+			if (!(found & HASH)) {
+				*hash_id = params[n].content.value.a;
+				found |= HASH;
```
I assumed that was an error in the text. Why have a parameter that must always mean the same thing as the absence of the parameter?

jforissier replied:
Sounds reasonable to me. Perhaps GP wants to prevent the use of weak hashes, but in any case why prevent the use of algorithms stronger than SHA-256?

jenswi-linaro commented on lib/libutee/tee_api_objects.c:
```diff
@@ -306,8 +306,33 @@ void TEE_CloseObject(TEE_ObjectHandle object)
+
+TEE_Result __GP11_TEE_AllocateTransientObject(TEE_ObjectType objectType,
+					      uint32_t maxKeySize,
+					      TEE_ObjectHandle *object)
```
I guess you mean the other way around. I'll update.

jenswi-linaro commented on lib/libutee/tee_api.c:
```diff
@@ -580,22 +580,30 @@ void TEE_GetREETime(TEE_Time *time)
+
+	if (hint == TEE_USER_MEM_HINT_NO_FILL_ZERO ||
+	    hint == TEE_MALLOC_NO_FILL ||
+	    hint == (TEE_MALLOC_NO_FILL | TEE_MALLOC_NO_SHARE))
```
I'll change to switch case, refuse `TEE_MALLOC_NO_SHARE`, and panic on `TEE_MALLOC_NO_FILL` without `TEE_MALLOC_NO_SHARE`.

jenswi-linaro commented on lib/libutee/tee_api_objects.c:
```diff
@@ -451,7 +451,7 @@ TEE_Result __GP11_TEE_PopulateTransientObject(TEE_ObjectHandle object,
 
 void TEE_InitRefAttribute(TEE_Attribute *attr, uint32_t attributeID,
-			  const void *buffer, uint32_t length)
+			  const void *buffer, size_t length)
```
My first approach was to avoid wrappers for these simple cases, however, I discovered that it didn't work if the TA uses function pointers for certain things. So instead of trying to figure out which functions matter and which don't, I'm trying to do all of them as well as possible.

Another advantage of this approach is that it doesn't clutter the tee_internal_api.h file too much.

jenswi-linaro commented on lib/libutee/tee_api_objects.c:
```diff
@@ -812,7 +819,7 @@ TEE_Result TEE_StartPersistentObjectEnumerator(TEE_ObjectEnumHandle
 TEE_Result TEE_GetNextPersistentObject(TEE_ObjectEnumHandle objectEnumerator,
 				       TEE_ObjectInfo *objectInfo,
-				       void *objectID, uint32_t *objectIDLen)
+				       void *objectID, size_t *objectIDLen)
```
No, this can't be solved by casting, not for a 64-bit target.

jforissier replied:
> No, this can't be solved by casting, not for a 64-bit target.

Agreed, so a temporary variable is needed, isn't it? Consider a 64-bit TA built using the older devkit calling into the new libutee.so, 64 bits would get copied into a place where only 32 are expected.

jenswi-linaro commented:
update

jenswi-linaro commented:
Squashed and rebased.

jforissier commented:
@jenswi-linaro as far as I'm concerned I am fine reviewing in this PR, and unless many changes are requested I think it should remain manageable (each commit is small). I just need to take some time to review :wink: Will do today or tomorrow.

jforissier commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -263,27 +263,52 @@
+
 /*
- * Fix GP Internal Core API v1.1
+ * These are used in the OP-TEE abi, due to an inconsistency in the v1.1
```
s/abi/ABI/

jforissier reviewed (commented):
Reviewing "GP131: Add TEE_ALG_ECDH_DERIVE_SHARED_SECRET and TEE_ALG_ECDSA_SHA*"

One nit below and in the commit description please remove the colons in the second list. With that:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
Commit "GP131: add TEE_ALG_SHA3_* and TEE_ALG_SHAKE* defines":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
Commit "GP131: add two new API error codes":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
Commit "GP131: Add TEE_ATTR_X448_* defines":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
Reviewing commit " GP131: Update TEE_ObjectInfo".

There is a missing space in the commit description: ...`and TEE_GetNextPersistentObject())are updated`...

Regarding the handling of `gpd.ta.doesNotCloseHandleOnCorruptObject`.You seem to consider this boolean as a hint that the TEE gives to the TA, I understand it the other way around. The GP spec is surprisingly ambiguous actually, but all the other `gpd.ta` entries are clearly information passed from the TA to the TEE core (`singleInstance`, `stackSize`, `endian`...) so I think this one should work the same. Besides, TEE core properties are exported to TAs via `gpt.tee.*`.

Perhaps `doesNotCloseHandleOnCorruptObject` should be introduced in a separate commit? In which case this one would be:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
Reviewing "GP131: Update TEE_Param" + "GP131: Update TEE_Param"

Regarding the comment about the `__weak` symbol not working as expected when libutee is a shared library: it is quite possible that ldelf has the symbol lookup rules not 100% right.

But I worry more about another thing. Shouldn't we consider `libutee.so` an ABI? Of course that would make things more complex. But if we don't, then it is impossible to run older TAs with newer libraries unless they are recompiled with the compat flag, doesn't it? Reason being: `lib/libutee/arch/arm/user_ta_entry.c` assumes a new-style `TA_OpenSessionEntryPoint()` for example. Which would be true for 1.3.1 TAs and newer TAs in 1.1 compat mode, but not for older TAs.

I believe we can support the above scenario with the help of weak external symbols in `libutee.so`. Such symbols resolve to NULL at runtime when no shared object in the executable provides an implementation. Here is an example.

```bash
#!/bin/bash

cat >foo.c <<_EOF_
#include <stdio.h>

void bar(void) __attribute__((weak));

void foo(void)
{
	if (bar)
		bar();
	else
		printf("No bar()\n");
}
_EOF_

cc -shared -o libfoo.so foo.c

cat >main.c <<_EOF_
#include <stdio.h>

void foo(void);

#ifdef WITH_BAR
void bar(void)
{
	printf("bar()\n");
}
#endif

int main(int argc, char *argv[])
{
	foo();
}
_EOF_

export LD_LIBRARY_PATH=.
cc -DWITH_BAR=1 -o main main.c -L. -lfoo
./main
cc -o main main.c -L. -lfoo
./main
```

Execution gives:
```
$ ./test.sh
bar()
No bar()
```

jenswi-linaro commented:
> Regarding the comment about the `__weak` symbol not working as expected when libutee is a shared library: it is quite possible that ldelf has the symbol lookup rules not 100% right.

The problem was while linking `os_test_lib` while using a static `libutee`, so `ldelf` hasn't entered the picture yet.

jenswi-linaro commented:
> But I worry more about another thing. Shouldn't we consider `libutee.so` an ABI? Of course that would make things more complex. But if we don't, then it is impossible to run older TAs with newer libraries unless they are recompiled with the compat flag, doesn't it? Reason being: `lib/libutee/arch/arm/user_ta_entry.c` assumes a new-style `TA_OpenSessionEntryPoint()` for example. Which would be true for 1.3.1 TAs and newer TAs in 1.1 compat mode, but not for older TAs.

Older TAs will not work with a newer dynamically linked `libutee.so` in the approach I've taken, it's not just about what I'm doing with `__ta_open_sess()` etc. Functions which has a new signature get a new `__GP11_` prefixed copy with the old signature. Old TAs are tricked using macros to use the `__GP11_` prefixed functions where available. To preserve the `libutee.so` ABI we would need to do it the other way around by adding a `__GP131_` prefix for the changed functions + macros instead.

If we could let old TA continue to use their old version of `libutee.so` and have new TAs use the new `libutee.so` we should, on the other hand, be OK.

jenswi-linaro commented:
> Then maybe it's time to look into symbol versioning and what ELF has to propose in this regard?

This seems quite powerful and may do exactly what we need. Do you have an experience with it?

jforissier commented:
> > Then maybe it's time to look into symbol versioning and what ELF has to propose in this regard?
> 
> This seems quite powerful and may do exactly what we need. Do you have an experience with it?

Unfortunately not.

jenswi-linaro commented:
There's some good information on symbol versioning at https://akkadia.org/drepper/dsohowto.pdf

Ldelf will need to be updated to take advantage of this. It might be a bit much to do at the moment. However, that paper also mentions another more coarse-grained method to preserve the ABI, simply renaming the library. By giving the shared library libutee a new UUID old TAs will not use the new library. If you have an old TA you just need to preserve the old libutee in the system when upgrading other parts.

So to summarize, we give a new UUID to libutee and old TAs can continue to work with old ta-dev-kits. Once ldelf supports versioned symbols we can use a more advanced way of maintaining a stable ABI.

jforissier commented:
> There's some good information on symbol versioning at https://akkadia.org/drepper/dsohowto.pdf

https://maskray.me/blog/2020-11-26-all-about-symbol-versioning has lots of details too (I have only skimmed through it)

> Ldelf will need to be updated to take advantage of this. It might be a bit much to do at the moment. 

Agreed.

> However, that paper also mentions another more coarse-grained method to preserve the ABI, simply renaming the library. By giving the shared library libutee a new UUID old TAs will not use the new library. If you have an old TA you just need to preserve the old libutee in the system when upgrading other parts.
> 
> So to summarize, we give a new UUID to libutee and old TAs can continue to work with old ta-dev-kits. Once ldelf supports versioned symbols we can use a more advanced way of maintaining a stable ABI.

Sounds like a very good option. In the mean time I would be quite interested in investigating symbol versioning support.

jenswi-linaro commented:
> Sounds like a very good option. In the mean time I would be quite interested in investigating symbol versioning support.

To link shared libraries with version script you can use this RFC PR https://github.com/OP-TEE/optee_os/pull/5704

jforissier reviewed (commented):
Reviewing "GP131: Update TEE_Whence"

s/TEE_Param/TEE_Whence/ in the commit description, then:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_OperationMode":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on lib/libutee/include/user_ta_header.h:
```diff
@@ -125,6 +125,7 @@ enum user_ta_prop_type {
 	USER_TA_PROP_TYPE_IDENTITY,	/* TEE_Identity */
 	USER_TA_PROP_TYPE_STRING,	/* zero terminated string of char */
 	USER_TA_PROP_TYPE_BINARY_BLOCK,	/* zero terminated base64 coded string */
+	USER_TA_PROP_TYPE_U64,	/* uint64 */
```
uint64_t

jforissier reviewed (commented):
For "GP131: Update TEE_GetProperty*() functions", one minor comment.

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_CheckMemoryAccessRights()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_Malloc() and TEE_Realloc()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_MemMove(), TEE_MemCompare() and TEE_MemFill()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_GetObjectBufferAttribute()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_AllocateTransientObject()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_InitRefAttribute()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_OpenPersistentObject()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on lib/libutee/include/tee_internal_api.h:
```diff
@@ -243,7 +243,10 @@ TEE_Result TEE_CloseAndDeletePersistentObject1(TEE_ObjectHandle object);
+				      size_t newObjectIDLen);
+TEE_Result __GP11_TEE_RenamePersistentObject(TEE_ObjectHandle object,
+					     const void *newObjectID,
+					     uint32_t newObjectIDLen);
```
For `TEE_InitRefAttribute()` you did not introduce a `_GP11_` function, but here you do. Why?

jenswi-linaro replied:
Because it was already introduced in "GP131: Update TEE_Attribute"

jforissier reviewed (commented):
Reviewing "GP131: Update TEE_RenamePersistentObject() and TEE_GetNextPersistentObjects()"

jforissier reviewed (commented):
For "GP131: Update TEE_ReadObjectData() and TEE_WriteObjectData()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_TruncateObjectData()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_SeekObjectData()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_GetOperationInfoMultiple()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_AE* functions":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on core/tee/tee_svc_cryp.c:
```diff
@@ -2192,7 +2192,14 @@ tee_svc_obj_ed25519_parse_params(const TEE_Attribute *params, size_t num_params,
+			else if (params[n].content.value.a == 1)
+				*ph_flag = true;
+			else
+				return TEE_ERROR_BAD_PARAMETERS;
```
I suppose this should ideally have been introduced in commit 0aaad418ac8b ("core: crypto: add Ed25519 support") and would justify a separate commit.

Looks OK with the spec and with xtest since commit [7596bd335890](https://github.com/OP-TEE/optee_test/commit/7596bd335890) ("regression 4016: fix TEE_ATTR_EDDSA_PREHASH parameter") but CC @varder in case this could cause compatibility troubles.

jenswi-linaro replied:
OK, done in https://github.com/OP-TEE/optee_os/pull/5707

jforissier reviewed (commented):
Reviewing "GP131: Update TEE_AsymmetricSignDigest() and TEE_AsymmetricVerifyDigest()"

The change without the `TEE_ATTR_EDDSA_PREHASH` fix is `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`.

jforissier reviewed (commented):
For "GP131: Update TEE_GenerateRandom()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

(although the change is not strictly needed, assuming the conversion is done systematically)

jforissier reviewed (commented):
For "GP131: Update TEE_BigInt*() functions":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: add ta property gpd.tee.internalCore.version":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: add TA property gpd.ta.endian":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: add TA property gpd.client.endian":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update gpd.tee.trustedStorage.antiRollback.protectionLevel":

s/root is can/root can/ then"

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For "GP131: Update TEE_FreeOperation()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
> For "GP131: Update TEE_GenerateRandom()":
> 
> `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
> 
> (although the change is not strictly needed, assuming the conversion is done systematically)

Yes, if this had been the only change needed in this class of changes I would have let it slide. However, with some functions, we initialize function pointer, and then we cannot have discrepancies like this. For instance `__GP11_TEE_DigestUpdate()` fixes an actual compile error in one of our test TAs. That's why I decided on a more strict update of the API. I'm still cheating with `TEE_SeekObjectData()`, but I find it unlikely that should cause problems. However, if you think it's an unnecessary risk I can fix that one too.

jenswi-linaro commented:
Rebased on top of https://github.com/OP-TEE/optee_os/pull/5705.
So please ignore "libutee: clarify that the used version is still 1.1" in 
this PR.

Dropped the gpd.ta.doesNotCloseHandleOnCorruptObject part from "GP131: 
Update TEE_ObjectInfo", will take that later.
Extracted what's now in https://github.com/OP-TEE/optee_os/pull/5707
from "GP131: Update TEE_AsymmetricSignDigest() and TEE_AsymmetricVerifyDigest()"

Squashed in update the "GP131: Update TEE_Param" and updated the commit
message as needed.

Added "Update reported version to GP TEE Core Internal API v1.3.1".

Tags applied.

etienne-lms commented on lib/libutee/tee_api_arith_mpi.c:
```diff
@@ -785,28 +821,56 @@ int32_t TEE_BigIntIsProbablePrime(const TEE_BigInt *op,
+				     size_t len __unused,
+				     const TEE_BigInt *modulus __unused)
+{
+	return TEE_SUCCESS;
```
("GP131: Update TEE_BigInt*() functions") `TEE_ERROR_NOT_IMPLEMENTED`?

jenswi-linaro replied:
No, this is intended behaviour.

jenswi-linaro replied:
It's an opportunity to do fast modular multiplication, but we've never bothered to do the fast part of FMM so we just use the regular bigint function. There is a comment a bit further up in this file that explains a bit. I'd also recommend looking in the specification.

etienne-lms replied:
Ok, right. thanks. 
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit "GP131: Update TEE_BigInt*() functions".

etienne-lms reviewed (commented):
A question on commit "GP131: Update TEE_BigInt*() functions" regarding `TEE_BigIntInitFMMContext1()` return value.

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with all commits with some minor comments (almost only missing local variables init value) for the 36 other commits.

jenswi-linaro commented on lib/libutee/tee_api_objects.c:
```diff
@@ -364,6 +382,35 @@ TEE_Result TEE_PopulateTransientObject(TEE_ObjectHandle object,
+					      const __GP11_TEE_Attribute *attrs,
+					      uint32_t attrCount)
+{
+	struct utee_attribute ua[attrCount];
```
This results in a compiler error:
```
lib/libutee/tee_api_objects.c: In function ‘__GP11_TEE_PopulateTransientObject’:
lib/libutee/tee_api_objects.c:389:9: error: variable-sized object may not be initialized
  389 |  struct utee_attribute ua[attrCount] = { };
      |         ^~~~~~~~~~~~~~
```
It is possible to use `memset()`, but I'm not sure how useful that is as we can depend on `__utee_from_gp11_attr()` to initialize `ua`. If it's OK with you I'd rather keep `ua` as it is, but I'll initialize all the other locals.

jenswi-linaro replied:
I'll wait with any updates until you've replied since it's messy enough to update all this one time.

etienne-lms replied:
ok, i'm fine as-is. indeed `__utee_from_gp11_attr()` initializes the full array.

jenswi-linaro commented on lib/libutee/tee_api_property.c:
```diff
@@ -341,16 +359,42 @@ TEE_Result TEE_GetPropertyAsU32(TEE_PropSetHandle propsetOrEnumerator,
+				    uint32_t *value_len)
+{
+	TEE_Result res;
+	enum user_ta_prop_type type;
```
This is probably in commit "GP131: Update TEE_GetProperty*() functions", I'll update in that commit.

jenswi-linaro commented:
Addressed comments add added tags for commented commits, rebased on master.

jenswi-linaro commented:
The following commits haven't been reviewed yet:
- GP131: Update TEE_RenamePersistentObject() and TEE_GetNextPersistentObject()
- Update reported version to GP TEE Core Internal API v1.3.1

jforissier commented:
For "Update reported version to GP TEE Core Internal API v1.3.1":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
For "GP131: Update TEE_RenamePersistentObject() and TEE_GetNextPersistentObject()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Rebased and tags applied.

jenswi-linaro commented:
All the previous commits are now reviewed. I'm adding a few more commit to 
update to version 1.3.1. I don't plan to add any more new commits to after
this.

Please ignore the following commits from #5720
- "core: syscall_storage_obj_write(): handle corrupt object"
- "libutils: use raw_malloc_*() as more primitive bget wrappers"
- "libutils: fix gen_malloc_buffer_overlaps_heap()"
and
- "core: crypto: fix TEE_ATTR_EDDSA_PREHASH interpretation"
from #5707

I've added support for SHA-3 and HMAC based on SHA-3. SHA-3 brings a new
kind of hash function with SHAKE128 and SHAKE256 (see "GP131: add
TEE_DigestExtract()".

There are still a few mandatory algorithms missing with SHA-3 and DSA, RSA, 
ECDH and ECDSA. But that might be a bit much to do here, especially since
test vectors for at least DSA is hard to come by.

jenswi-linaro commented:
The rust stuff is a bit frustrating. I can't see what's wrong in the log except that it times out. When trying to build locally the binaries fail to execute with:
```
# hello_world-rs 
hello_world-rs: /lib64/libc.so.6: version `GLIBC_2.33' not found (required by hello_world-rs)
hello_world-rs: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by hello_world-rs)
hello_world-rs: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by hello_world-rs)
```
I guess that's because the rust compiler ships with a newer libc than what we have in the gcc cross compiler. But I haven't been able to find it yet. It seems that the Teaclave TrustZone SDK reaches into the internals of libutee which makes it quite hard to do any significant update. Perhaps it's best to ignore rust in the PR and ask the teaclave people to fix whatever needs to be fixed once it's merged.

jforissier commented:
> The rust stuff is a bit frustrating. I can't see what's wrong in the log except that it times out.

The secure and non-secure console output is not shown on failure and perhaps we should change that. That being said I don't think any issue can be fixed without reproducing it in a dev environment anyways so not sure it matters that much.

> When trying to build locally the binaries fail to execute with:
> 
> ```
> # hello_world-rs 
> hello_world-rs: /lib64/libc.so.6: version `GLIBC_2.33' not found (required by hello_world-rs)
> hello_world-rs: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by hello_world-rs)
> hello_world-rs: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by hello_world-rs)
> ```
> 
> I guess that's because the rust compiler ships with a newer libc than what we have in the gcc cross compiler. But I haven't been able to find it yet.

Please see below for how to reproduce the issue with the Docker image used in CI. The TA panics, but the Expect script (`optee_rust/ci/qemu-check.exp`) doesn't detect this failure mode so it just times out (this could/should probably be improved).

> It seems that the Teaclave TrustZone SDK reaches into the internals of libutee which makes it quite hard to do any significant update.  Perhaps it's best to ignore rust in the PR and ask the teaclave people to fix whatever needs to be fixed once it's merged.

I'd rather have any issue discussed here before the PR is merged, because knowingly breaking things without a clear plan on a fix could cause some pain downstream.

```
$ docker run -it --rm jforissier/optee_os_ci:qemuv8_check
root@f7a555aac192:~# cd ~/optee_repo_qemu_v8/optee_os/
root@f7a555aac192:~/optee_repo_qemu_v8/optee_os# git fetch github pull/5688/merge
root@f7a555aac192:~/optee_repo_qemu_v8/optee_os# git checkout FETCH_HEAD
root@f7a555aac192:~/optee_repo_qemu_v8/optee_os# cd ../build
root@f7a555aac192:~/optee_repo_qemu_v8/build# make -j$(nproc) OPTEE_RUST_ENABLE=y check-rust
[...]
Starting QEMU...
 done, guest is booted.
Test Rust applications:
Running acipher-rs...
[hangs here]
^C

root@f7a555aac192:~/optee_repo_qemu_v8/build# cd ../out/bin
root@f7a555aac192:~/optee_repo_qemu_v8/out/bin# tail  serial0.log 
Starting network: OK
Starting network (udhcpc): OK
Starting domain watchdog daemon: xenwatchdogd startup


Welcome to Buildroot, type root or test to login
buildroot login: root
# acipher-rs 256 teststring
Error: Trusted Application has panicked during the operation. (error code 0xffff3024)


root@f7a555aac192:~/optee_repo_qemu_v8/out/bin# tail -30  serial1.log 
D/TC:? 0 ldelf_syscall_open_bin:146 res=0
D/LD:  ldelf:168 ELF (057f4b66-bdab-11eb-96cf-33d6e41cc849) at 0x8008d000
[+] TA create
[+] TA open session
[+] TA invoke command
[+] TA invoke command
[+] TA invoke command
E/TC:? 0 
E/TC:? 0 TA panicked with code 0xffff0001
E/LD:  Status of TA 057f4b66-bdab-11eb-96cf-33d6e41cc849
E/LD:   arch: aarch64
E/LD:  region  0: va 0x80004000 pa 0x0e32a000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x80006000 pa 0x0e32c000 size 0x008000 flags r-xs (ldelf)
E/LD:  region  2: va 0x8000e000 pa 0x0e334000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x8000f000 pa 0x0e335000 size 0x004000 flags rw-s (ldelf)
E/LD:  region  4: va 0x80013000 pa 0x0e339000 size 0x001000 flags r--s
E/LD:  region  5: va 0x80014000 pa 0x0e379000 size 0x001000 flags rw-s (stack)
E/LD:  region  6: va 0x80015000 pa 0x773a0ad0 size 0x001000 flags rw-- (param)
E/LD:  region  7: va 0x80016000 pa 0x773a0a60 size 0x001000 flags rw-- (param)
E/LD:  region  8: va 0x8008d000 pa 0x00010000 size 0x032000 flags r-xs [0]
E/LD:  region  9: va 0x800bf000 pa 0x00042000 size 0x00d000 flags rw-s [0]
E/LD:   [0] 057f4b66-bdab-11eb-96cf-33d6e41cc849 @ 0x8008d000
E/LD:  Call stack:
E/LD:   0x800a7ccc
E/LD:   0x80096108
D/TC:? 0 user_ta_enter:176 tee_user_ta_enter: TA panicked with code 0xffff0001
D/TC:? 0 destroy_ta_ctx_from_session:341 Remove references to context (0x7741bad8)
D/TC:? 0 destroy_context:326 Destroy TA ctx (0x7741bac0)
D/TC:? 0 tee_ta_close_session:529 csess 0x7741bb20 id 2
D/TC:? 0 tee_ta_close_session:548 Destroy session
```

jenswi-linaro commented:
Thanks for the instructions on how to build and test the rust examples. The rust wrappers ignore the `tee_internal_api.h` file and instead assume an ABI declared in `optee-utee-sys/src/tee_api.rs` independent of what is in the libutee in use. This is similar to the problem with TAs linked against libutee.so except that instead of establishing the dependency at link time it's established manually in `optee-utee-sys/src/tee_api.rs` so it cannot be updated automatically by recompiling for instance.

Until `optee-utee-sys/src/tee_api.rs` is updated it must stick to a `libtee` which provides the ABI described in `optee-utee-sys/src/tee_api.rs`. With this approach in rust when linking against non-rust libraries the same problem must be faced again and again with for instance libc.

We need feedback from the rust maintainer on how to proceed.

jenswi-linaro commented:
Rebased to keep up to date.

jenswi-linaro commented:
> Thanks for the instructions on how to build and test the rust examples. The rust wrappers ignore the `tee_internal_api.h` file and instead assume an ABI declared in `optee-utee-sys/src/tee_api.rs` independent of what is in the libutee in use. This is similar to the problem with TAs linked against libutee.so except that instead of establishing the dependency at link time it's established manually in `optee-utee-sys/src/tee_api.rs` so it cannot be updated automatically by recompiling for instance.
> 
> Until `optee-utee-sys/src/tee_api.rs` is updated it must stick to a `libtee` which provides the ABI described in `optee-utee-sys/src/tee_api.rs`. With this approach in rust when linking against non-rust libraries the same problem must be faced again and again with for instance libc.
> 
> We need feedback from the rust maintainer on how to proceed.

@fanrong1992 @mssun @DemesneGH do you have any thoughts on how to fix the Teaclave problems when updating the TEE Internal Core API?

DemesneGH commented:
Hi @jenswi-linaro , I think we can generate the `tee_api.rs` wrappers automatically using `bindgen`, but it also needs some manual work on adjusting the caller's parameters. After the GP 1.3.1 upgrade is finished please let me know then I will start to update the Teaclave TrustZone SDK. Thanks!

jenswi-linaro commented:
Hi @DemesneGH, thanks. We plan to merge this after the next release at the end of this month if the review is ready by then. I don't plan to add any more commits to this pull request, but may need do some small changes during the review. This should cover all the tricky things with regard to compatibility. Some more algorithms will be added at some later stage but that's proper extensions and shouldn't break anything.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"Update reported version to GP TEE Core Internal API v1.3.1"
"GP131: Add TEE_ALG_ECDH_DERIVE_SHARED_SECRET and TEE_ALG_ECDSA_SHA*"
"GP131: add TEE_ALG_SHA3_* and TEE_ALG_SHAKE* defines"
"GP131: add two new API error codes"
"GP131: Add TEE_ATTR_X448_* defines"
"GP131: Update HKDF defines"
"GP131: Update TEE_ObjectInfo"
"GP131: Update TEE_Attribute"
"GP131: Update TEE_Param" (already tagged)
"GP131: Update TEE_Whence"
"GP131: Update TEE_OperationMode" (already tagged)
"GP131: Update TEE_GetProperty*() functions"
"GP131: Update TEE_CheckMemoryAccessRights()"
"GP131: Update TEE_Malloc() and TEE_Realloc()"
"GP131: Update TEE_MemMove(), TEE_MemCompare() and TEE_MemFill()"
"GP131: Update TEE_GetObjectBufferAttribute()"
"GP131: Update TEE_AllocateTransientObject()"
"GP131: Update TEE_InitRefAttribute()"
"GP131: Update TEE_OpenPersistentObject()"
"GP131: Update TEE_CreatePersistentObject()"
"GP131: Update TEE_RenamePersistentObject() and TEE_GetNextPersistentObject()"
"GP131: Update TEE_ReadObjectData() and TEE_WriteObjectData()"
"GP131: Update TEE_TruncateObjectData()"
"GP131: Update TEE_GetOperationInfoMultiple()"
"GP131: Update TEE_{Digest,Cipher,MAC}* functions" **(with comment addressed)**
"GP131: Update TEE_AE* functions" **(with comment addressed)**

Another comment related to that last commit ("Update TEE_AE*") but can be addressed reperatly: Looking at the spec (v1.0 to v1.3), `destLen` can be NULL ("omitted") in `TEE_AEEncryptFinal()` and `TEE_AEDecryptFinal()`. Current implementation does not allow that.


`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit:
"GP131: Update TEE_AsymmetricEncrypt() and TEE_AsymmetricDecrypt()"

For commit "GP131: Update TEE_AsymmetricSignDigest()
and TEE_AsymmetricVerifyDigest()":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with commit log fixed. It says _"For TEE_ALG_ED25519 use a more strict interpretation of_ TEE_ATTR_EDDSA_PREHASH as described in the specification.". This is true but not related to that commit and already stated in merged commit "core: crypto: fix TEE_ATTR_EDDSA_PREHASH interpretation".


`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"GP131: Update TEE_GenerateRandom()"
"GP131: Update TEE_BigInt*() functions" (already tagged)
"GP131: add ta property gpd.tee.internalCore.version" 
"GP131: add TA property gpd.ta.endian"
"GP131: add TA property gpd.client.endian"
"GP131: Update gpd.tee.trustedStorage.antiRollback.protectionLevel"
"GP131: Update TEE_FreeOperation()"

**Note on "gpd.tee.*version":** OP-TEE still implements "gpd.tee.apiversion" in a non conformant way (default "GPD-X.Y-dev", configurable with from CFG_TEE_API_VERSION). I think we should remove both CFG_TEE_API_VERSION and gpd.tee.apiversion.

Comment for comit "GP131: add missing TEE_ALG_* and TEE_TYPE_* defines": issue with `TEE_ALG_SM2_PKE`.

For commit "core: crypto: add API for SHA-3"
and commit "core: LTC: conditionally enable SHA-3 implementation":
LGTM (minor comments) but the commit message. First commit mentions "Adds LTC fallback for mbedtls configurations since mbedtls doesn't support SHA-3 yet.". This should be in the message of 2nd commit which could also mention that it also fixes declarations of existing crypto_hmac_*_alloc_ctx() functions according to embedded algorithms (config switches).
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` once addressed.

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit:
"Enable SHA-3 support by default"
"core: syscall_hash_final() support XOF"


Comments for commit "GP131: add TEE_DigestExtract()".
Comments for commit "GP131: Update TEE_CheckMemoryAccessRights()".


`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"GP131: check permissions on buffer passed to other TAs"


For commit "GP131: add support for TEE_MALLOC_NO_SHARE":
Implementation LGTM for the commit and its rust fixup commit. My question relates to `TEE_USER_MEM_HINT_NO_FILL_ZERO` that should relate only to a NO_SHARE|NO_FILL hint. The specs says "When allocating to a TA, a Trusted OS SHALL zero fill memory which (...) * Does not have the TEE_MALLOC_NO_SHARE hint.". Or do we want to allow our extension hint on sharable memory? 

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"Upgrade in-tree TAs to use latest version of TEE Internal Core API"

For commit "GP131: Update TEE_SetOperationKey*()"
Since v1.2 `TEE_SetOperationKey2()`  shall return `TEE_ERROR_SECURITY` when key1 and key2 handles are the same. Before v1.2 "_implementation MAY allow key1 and key2 to be the same._". OP-TEE currently does allow same keys.


`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"GP131: add TEE_BigIntAssign()"
"GP131: add TEE_BigIntSetBit()"
"GP131: add TEE_BigIntAbs()"
"GP131: add TEE_BigIntExpMod()" **(with comment addressed)**
"GP131: TEE_Asymmetric{En,De}crypt() add return codes"
"GP131: Add TA property gpd.ta.doesNotCloseHandleOnCorruptObject"

jenswi-linaro commented on lib/libutee/tee_api_operations.c:
```diff
@@ -1195,9 +1235,24 @@ TEE_Result TEE_CipherDoFinal(TEE_OperationHandle operation,
+
+	if (destLen)
+		__utee_check_inout_annotation(destLen, sizeof(*destLen));
+	dl = *destLen;
```
This relates to the "destLen can be NULL" issue we have also with the TEE_AE* functions.

I propose to treat a NULL destLen as if *destLen was 0, as we're doing today.
Then we'll address the "destLen can be NULL" issue at a later stage. Depending on how it's approached it might not be an easy task. I've already done one halfhearted attempt and it wasn't pretty.

etienne-lms replied:
I meant:
```c
	if (destLen) {
		__utee_check_inout_annotation(destLen, sizeof(*destLen));
		dl = *destLen;
	}
	res = TEE_CipherDoFinal(operation, srcData, srcLen, destData, &dl);
	if (destLen)
		*destLen = dl;

jenswi-linaro replied:
OK, I'll add that.

jenswi-linaro commented on lib/libutee/tee_api_operations.c:
```diff
@@ -1535,37 +1549,103 @@ TEE_Result TEE_AEUpdate(TEE_OperationHandle operation, const void *srcData,
 	if (res != TEE_SUCCESS)
 		goto out;
 
+	if (res != TEE_ERROR_SHORT_BUFFER && srcLen)
```
Thanks, I'll update.

jenswi-linaro replied:
No, that's not right. It's the `if (res != TEE_SUCCESS)` check that should be removed. I'll make it so instead.

etienne-lms replied:
maybe:
```c
	res = ae_update_helper(operation, srcData, srcLen, destData, destLen);
	if (res == TEE_SUCCESS && srcLen)
		operation->operationState = TEE_OPERATION_STATE_ACTIVE;

out:
```

jenswi-linaro commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -217,6 +218,7 @@
 #define TEE_OPERATION_KEY_DERIVATION       8
 #define TEE_OPERATION_STATE_INITIAL        0x00000000
 #define TEE_OPERATION_STATE_ACTIVE         0x00000001
+#define TEE_OPERATION_STATE_EXTRACTING     0x00000003
```
Thanks, I'll fix.

jenswi-linaro commented on lib/libutee/tee_api.c:
```diff
@@ -223,6 +223,53 @@ static void update_out_gp11_param(__GP11_TEE_Param params[TEE_NUM_PARAMS],
+	vaddr_t b1 = (vaddr_t)buf1;
+	vaddr_t b2 = (vaddr_t)buf2;
+	vaddr_t e1 = b1 + sz1 - 1;
+	vaddr_t e2 = b2 + sz2 - 1;
```
That's not supposed to be possible. I'll add a comment in TEE_CheckMemoryAccessRights().

jenswi-linaro commented on lib/libutee/tee_api_arith_mpi.c:
```diff
@@ -680,6 +680,52 @@ static bool mpi_is_even(mbedtls_mpi *x)
+	else
+		get_mpi(&mpi_op2, op2);
+
+	if (mbedtls_mpi_cmp_int(&mpi_n, 2) < 0)
```
Thanks, I'll fix.

jenswi-linaro commented:
Thanks for the review, I've address comments and added the tags except for:
- "GP131: Update TEE_{Digest,Cipher,MAC}* functions"
  Same as with the TEE_AE* functions below. OK to fix the "destLen can be
  NULL" issue later?
- "GP131: Update TEE_AE* functions"
  Yes, I agree that it's preferrable to take the "destLen can be NULL"
  change separately. Depending on how it's approached it might not be an
  easy task. Did you have any more comments on that commit?

I've updated "GP131: add TEE_DigestExtract()" as requested.

I've added a comment for "GP131: Update TEE_CheckMemoryAccessRights()".

Regarding "gpd.tee.version", it's deprecated as you say and possibly not
correctly implemented either. However, we've had it there for some time and
it hasn't caused any known issues so I'd rather remove that at a later
stage since it doesn't have much to do with updating to v1.3.1. Besides
decprecated is just a warning that it will be removed at some later point.

How should we deal with TEE_ALG_SM2_PKE? It was defined as 0x80000045 in
the v1.2 spec and is now defined as 0x80000046 in the v1.3.1 spec.
TEE_ALG_SM2_PKE was redefined in v1.3 when TEE_ALG_X448 was introduced as
0x80000045. Until now I haven't dealt with conflicts where a value has been
reused. I don't have any good workaround for this. I'd like update the
values without any attempt at backwards compatibility, the alternative with
conditional (depding on how the TA was compiled) translation of the value
in OP-TEE Core is not appealing. Besides it would still not cover how we
test stuff in xtest.

Regarding TEE_USER_MEM_HINT_NO_FILL_ZERO and sharable memory. I've
understood the difference between sharable non-sharable memory that it's
only the content of the memory not the physical memory. OP-TEE always uses
a temporary bounce buffer when calling a TA from another TA (except in the
PTA case, but that doesn't apply here). So the difference becomes that
OP-TEE refuses to copy from non-sharable memory into this bounce buffer.
Non-sharable memory serves as an extra protection layer against
the unintentional sharing of secrets. TEE_USER_MEM_HINT_NO_FILL_ZERO is still
useful, but only for sharable memory since non-sharable memory already have
TEE_MALLOC_NO_FILL.

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"GP131: Update TEE_SeekObjectData()"
"GP131: Update TEE_SetOperationKey*()" (with its fixup commit)

> "GP131: Update TEE_{Digest,Cipher,MAC}* functions"
Same as with the TEE_AE* functions below. OK to fix the "destLen can be
NULL" issue later?

Ok to address it later.

> "GP131: Update TEE_AE* functions"
Yes, I agree that it's preferrable to take the "destLen can be NULL"
change separately. Depending on how it's approached it might not be an
easy task. Did you have any more comments on that commit?

No more comments. Ok to address NULL destLen seperatly.
Change LGTM with the fixup commit. My R-b applies.

> I've updated "GP131: add TEE_DigestExtract()" as requested.
> I've added a comment for "GP131: Update TEE_CheckMemoryAccessRights()".

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for these 2 commits (with fixup commit foir the later).

> How should we deal with TEE_ALG_SM2_PKE? 

(related commit "GP131: add missing TEE_ALG_* and TEE_TYPE_* defines")

I would tend to use the v1.3 values straight, without any attempt at backwards compatibility as you said. Maybe time for a 4.x.0 tag series if we break backward compat. FWIW.


> Regarding TEE_USER_MEM_HINT_NO_FILL_ZERO and sharable memory...

(related commit "GP131: add support for TEE_MALLOC_NO_SHARE")

I mentionned the spec sentence _"When allocating to a TA, a Trusted OS SHALL zero fill memory which (...) "_  since it looks like a strong directive from the spec. I've understood it as when allocating memory that can be shared with another TA, it must be wiped from its previous content unless what old content (even from the very same TA) may leak sensitive information when the buffer or part of is shared.

jenswi-linaro commented:
Tags applied.

> > How should we deal with TEE_ALG_SM2_PKE?
> 
> (related commit "GP131: add missing TEE_ALG_* and TEE_TYPE_* defines")
> 
> I would tend to use the v1.3 values straight, without any attempt at backwards compatibility as you said. Maybe time for a 4.x.0 tag series if we break backward compat. FWIW.

If we're going to update major version then we should plan to get rid of other annoyances too.

> 
> > Regarding TEE_USER_MEM_HINT_NO_FILL_ZERO and sharable memory...
> 
> (related commit "GP131: add support for TEE_MALLOC_NO_SHARE")
> 
> I mentionned the spec sentence _"When allocating to a TA, a Trusted OS SHALL zero fill memory which (...) "_ since it looks like a strong directive from the spec. I've understood it as when allocating memory that can be shared with another TA, it must be wiped from its previous content unless what old content (even from the very same TA) may leak sensitive information when the buffer or part of is shared.

It is cleared already when the TA is initialized so that part shouldn't be a problem. If `TEE_USER_MEM_HINT_NO_FILL_ZERO` is used one must assume that the user understands that there might be sensitive garbage (from the current TA) in the allocated buffer.

etienne-lms commented:
>>> How should we deal with TEE_ALG_SM2_PKE?
>>
>> I would tend to use the v1.3 values straight, without any attempt at backwards compatibility as you said. Maybe time for a 4.x.0 tag series if we break backward compat. FWIW.
>
>If we're going to update major version then we should plan to get rid of other annoyances too.

+1 for the 2.x and very early 3.x compat things.


>> Regarding TEE_USER_MEM_HINT_NO_FILL_ZERO and sharable memory...
>>
>> I mentionned the spec sentence "When allocating to a TA, a Trusted OS SHALL zero fill memory which (...) " since it looks like a strong directive from the spec. I've understood it as when allocating memory that can be shared with another TA, it must be wiped from its previous content unless what old content (even from the very same TA) may leak sensitive information when the buffer or part of is shared.
>
> It is cleared already when the TA is initialized so that part shouldn't be a problem. If TEE_USER_MEM_HINT_NO_FILL_ZERO is used one must assume that the user understands that there might be sensitive garbage (from the current TA) in the allocated buffer.

I would think directive "... OS SHALL zero fill ..." is there is protect the TA against itself. That said, the hint being an OP-TEE extension, one using it should indeed knows what it implies.

Ok as is. 
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`- for commit
"GP131: add support for TEE_MALLOC_NO_SHARE"

jenswi-linaro commented:
> > > > How should we deal with TEE_ALG_SM2_PKE?
> > > 
> > > 
> > > I would tend to use the v1.3 values straight, without any attempt at backwards compatibility as you said. Maybe time for a 4.x.0 tag series if we break backward compat. FWIW.
> > 
> > 
> > If we're going to update major version then we should plan to get rid of other annoyances too.
> 
> +1 for the 2.x and very early 3.x compat things.

I'm sorry, what do you mean?

jenswi-linaro commented:
Rebased on master. There's one commit left to review:
- GP131: add missing TEE_ALG_* and TEE_TYPE_* defines

jforissier commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -355,6 +383,10 @@
 #define TEE_TYPE_HMAC_SHA384                0xA0000005
 #define TEE_TYPE_HMAC_SHA512                0xA0000006
 #define TEE_TYPE_HMAC_SM3                   0xA0000007 /* Not in spec */
+#define TEE_TYPE_HMAC_SHA3_224              0xA0000008
```
`TEE_TYPE_HMAC_SM3` is in the GP 1.3.1 spec so please remove the comment

jforissier reviewed (commented):
Commit "GP131: add missing TEE_ALG_* and TEE_TYPE_* defines"

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

How about removing the comments `/* vx.y.z spec */` in a separate commit? They are not super useful and are not applied systematically (especially with this new patch).

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit
"GP131: Update TEE_{Digest,Cipher,MAC}* functions".

jenswi-linaro commented:
Tags applied. Added "GP131: tee_api_defines.h: remove obsolete comments"

jforissier commented:
> Tags applied. Added "GP131: tee_api_defines.h: remove obsolete comments"

...which is:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -327,12 +347,20 @@
 
 #define TEE_ALG_ED25519                         0x70006043 /* v1.3.1 spec */
+#define TEE_ALG_ED448                           0x70006044
 #define TEE_ALG_SM2_PKE                         0x80000045
```
Fix `TEE_ALG_SM2_PKE` to `0x80000046`

jenswi-linaro replied:
I'm taking that in a new commit.

etienne-lms commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -327,12 +347,20 @@
 #define TEE_ALG_HKDF                            0x80000047
 #define TEE_ALG_SM3                             0x50000007
 #define TEE_ALG_X25519                          0x80000044
+#define TEE_ALG_X448                            0x80000045
```
Could we sort this block of macros in value order ? (from line 355 to line 363).
By the way, `TEE_ALG_ILLEGAL_VALUE` should move to after last algo ID, line 371.

jenswi-linaro replied:
If it should be reordered it should be to match the order in the specification, but I'd rather treat that as out of scope for this PR.

etienne-lms reviewed (changes requested):
For commit "GP131: add missing TEE_ALG_* and TEE_TYPE_* defines": `TEE_ALG_SM2_PKE` value needs to be fixed.

jenswi-linaro commented:
> Could we sort this block of macros in value order ? (from line 355 to line 363).
By the way, TEE_ALG_ILLEGAL_VALUE should move to after last algo ID, line 371.

If it should be reordered it should be to match the order in the specification, but I'd rather treat that as out of scope for this PR.

jenswi-linaro commented:
Tag applied, added "GP131: Fix the TEE_ALG_SM2_PKE define"

jforissier commented:
> Tag applied, added "GP131: Fix the TEE_ALG_SM2_PKE define"

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Tag applied.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"GP131: add missing TEE_ALG_* and TEE_TYPE_* defines",
"GP131: tee_api_defines.h: remove obsolete comments"
and "GP131: Fix the TEE_ALG_SM2_PKE define".

jenswi-linaro commented:
Tags applied

jforissier commented:
How should this be tested? By upgrading `xtest` TAs to 1.3.1 I suppose, while still keeping the/some older code to be used when `CFG_TA_OPTEE_CORE_API_COMPAT_1_1=y`?

jenswi-linaro commented:
> How should this be tested? By upgrading `xtest` TAs to 1.3.1 I suppose, while still keeping the/some older code to be used when `CFG_TA_OPTEE_CORE_API_COMPAT_1_1=y`?

That's done in https://github.com/OP-TEE/optee_test/pull/634

jforissier commented:
> > How should this be tested? By upgrading `xtest` TAs to 1.3.1 I suppose, while still keeping the/some older code to be used when `CFG_TA_OPTEE_CORE_API_COMPAT_1_1=y`?
> 
> That's done in [OP-TEE/optee_test#634](https://github.com/OP-TEE/optee_test/pull/634)

Ah, very good. How did I missed it? :open_mouth: So that PR upgrades to 1.3.1 but doesn't leave any way to test the compatibility flag. Does it mean we will handle this flag on a "best effort" basis, i.e., not tested by us and fixed whenever an issue is reported? That would be OK with me.

jenswi-linaro commented:
For now, it can be tested with the GP tests.

jforissier commented:
So I think this is ready for merging.

FYI, I have investigated a bit about symbol versioning for the shared libutee. Everything seems to behave as expected, I have a development branch here: https://github.com/jforissier/optee_os/commits/symver but of course it is useless until ldelf supports versioned symbol lookup.

jforissier commented:
@jenswi-linaro could you address the checkpatch issues? Thanks.

jenswi-linaro commented:
Fixed the checkpatch issues, except the ones about typedefs since I believe it's more consistent to use typedefs for the compatibility types.

jforissier commented:
I can see that the deprecated `TEE_ALG_ECDSA_P521` is still used in `core/drivers/crypto/*` and `lib/libmbedtls/core/ecc.c`. Shouldn't `TEE_ALG_ECDSA_SHA512` be used instead? And same for the other Ps.

jenswi-linaro commented:
> TEE_ALG_ECDSA_P521

Good point. I'd rather take that after this PR if that's OK since it's more about internal maintenance.

jforissier commented:
> > TEE_ALG_ECDSA_P521
> 
> Good point. I'd rather take that after this PR if that's OK since it's more about internal maintenance.

Fine with me.

jforissier commented:
Don't worry about the CI failure in "make check (QEMUv8)". I believe the "no space left" is caused by an update of the Docker image I did in preparation to build with BTI enabled (I added a toolchain). I am reverting this change now an will restart the job.

jbech-linaro commented:
> Don't worry about the CI failure

I noticed that different PRs have failed at the same location on IBART today. So, I get the impression that we have sneaked in something that either introduced a regression or alternatively we need to update compile time flags.

jenswi-linaro commented:
```
  4135:  2023-01-30 09:45:35 (873 KB/s) - ‘/dev/shm/qemu/out-br/build/.fakeroot_1.26.orig.tar.gz.l4aQO9/output’ saved [549685/549685]
  4136:  
  4137:  fakeroot_1.26.orig.tar.gz: OK (sha256: 480a578ffdc5592e73df4c086950d321b4adc78dbdaec56c82e4fe1fb68de8e9)
  4138:  mktemp: failed to create file via template ‘/mnt/storage/buildroot_cache/download/fakeroot/fakeroot_1.26.orig.tar.gz.XXXXXX’: Permission denied
  4139:  make[2]: *** [package/pkg-generic.mk:189: /dev/shm/qemu/out-br/build/host-fakeroot-1.26/.stamp_downloaded] Error 1
  4140:  make[1]: *** [Makefile:23: _all] Error 2
  4141:  make[1]: Leaving directory '/dev/shm/qemu/out-br'
  4142:  make: *** [common.mk:346: buildroot] Error 2
  4143:  make: *** Waiting for unfinished jobs....
```
Perhaps something in the file system on IBART?

jbech-linaro commented:
> Perhaps something in the file system on IBART?

Perhaps, I'll have a look after some meetings, thanks!

jforissier commented:
Now it's the GitHub CI that doesn't want to cooperate...
```
Complete job name: make check (QEMUv8)
...
2023-01-30T15:02:52.0241214Z /root/optee_repo_qemu_v8/out-br-domu/per-package/host-libopenssl/host/bin/ccache /usr/bin/gcc  -I. -Iinclude -fPIC -pthread -m64 -Wa,--noexecstack -O2 -I/root/optee_repo_qemu_v8/out-br-domu/per-package/host-libopenssl/host/include -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/root/optee_repo_qemu_v8/out-br-domu/per-package/host-libopenssl/host/etc/ssl\"" -DENGINESDIR="\"/root/optee_repo_qemu_v8/out-br-domu/per-package/host-libopenssl/host/lib64/engines-1.1\"" -DZLIB -DZLIB_SHARED -DNDEBUG -I/root/optee_repo_qemu_v8/out-br-domu/per-package/host-libopenssl/host/include -MMD -MF crypto/asn1/tasn_prn.d.tmp -MT crypto/asn1/tasn_prn.o -c -o crypto/asn1/tasn_prn.o crypto/asn1/tasn_prn.c
2023-01-30T15:02:52.1859506Z /usr/bin/ld: final link failed: No space left on device
2023-01-30T15:02:52.1872001Z collect2: error: ld returned 1 exit status
```
:cry:

etienne-lms commented:
Thanks for the heavy work :clap: 
This spec update is a great archievement
